include/lbann/utils/argument_parser.hpp

```diff
@@ -0,0 +1,586 @@
+   *  @return A const-reference to the held value.
+   */
+  template <typename T>
+  T const& get(std::string const& option_name) const;
+
+  ///@}
+  /** @name Output */
+  ///@{
+
+  /** @brief Print a help string to a stream.
+   *  @param[in] stream The ostream to print the help message to.
+   */
+  void print_help(std::ostream& stream) const;
+
+  ///@}
+
+private:
+
+  /** @brief Implementation of add_flag */
+  readonly_reference<bool>
+  add_flag_impl_(std::string const& name,
+                 std::initializer_list<std::string> cli_flags,
+                 std::string const& description,
+                 bool default_value);
+
+private:
+  /** @brief Dictionary of arguments to their values */
+  std::unordered_map<std::string, utils::any> params_;
+  /** @brief Patch around in-progress clara limitation */
+  std::unordered_set<std::string> required_;
+  /** @brief The underlying clara object */
+  clara::Parser parser_;
+  /** @brief The name of the executable. */
+  std::string exe_name_ = "<exe>";
+
+};
+
+inline bool
+argument_parser::option_is_defined(std::string const& option_name) const
+{
+  return params_.count(option_name);
+}
+
+template <typename T>
+inline T const& argument_parser::get(std::string const& option_name) const
+{
+  return utils::any_cast<T const&>(params_.at(option_name));
+}
+
+template <typename T>
+inline auto argument_parser::add_option(
+  std::string const& name,
+  std::initializer_list<std::string> cli_flags,
+  std::string const& description,
+  T default_value)
+  -> readonly_reference<T>
+{
+  params_[name] = std::move(default_value);
+  auto& param_ref = any_cast<T&>(params_[name]);
+  clara::Opt option(param_ref, name);
+  for (auto const& f : cli_flags)
+    option[f];
+  parser_ |= option(description).optional();
+  return param_ref;
+}
+
+template <typename T>
+inline size_t argument_parser::add_argument(
+  std::string const& name,
+  std::string const& description,
+  T default_value)
+{
+  params_[name] = std::move(default_value);
+  parser_ |= clara::Arg
+    (utils::any_cast<T&>(params_[name]), name)
+    (description).optional();
+  return parser_.m_args.size() - 1;
+}
+
+template <typename T>
+inline size_t argument_parser::add_required_argument(
+  std::string const& name,
+  std::string const& description)
+{
+  // Add the reference to bind to
+  params_[name] = T{};
+  auto& param_ref = params_[name];
+
+  required_.insert(name);
+
+  // Make sure the required arguments are all grouped together.
+  auto iter = parser_.m_args.cbegin(), invalid = parser_.m_args.cend();
+  while (iter != invalid && !iter->isOptional())
+    ++iter;
+
+  // Create the argument
+  auto ret = parser_.m_args.emplace(
+    iter,
+    [name,&param_ref,this](std::string const& value)
+    {
+      auto result = clara::detail::convertInto(
+        value, utils::any_cast<T&>(param_ref));
+      if (result)
+        required_.erase(name);
+      return result;
+    },
+    name);
```

Comment:
Does this properly handle the case where someone sets a required argument after an optional one? It looks like the index returned by `add_argument` will be invalid. Perhaps it would be better just to throw an exception if the user tries something weird. This also allows you to implement this similarly to `add_argument`.

Reply:
I will improve the documentation. See the `@return` for `add_argument` around line 346 of this file for the official position: Adding an "optional" argument will give you its current index in the parser at the time of writing. Adding a required argument after that will invalidate the optional parameter's returned index (predictably -- its new index will be the returned index plus one).

I was on the fence about what to return here. I don't know how relevant the index into the positional argument list actually is -- I've seen this behavior elsewhere but I've never encountered a real use-case. I could just have them return references to their members. (This is actually probably what I should do...) I could also have them return nothing.

I would also be hesitant to prohibit adding required arguments after adding optional arguments, given some of the (IMO, misguided and unnecessary) semantic requests for argument parsing. There's really nothing wrong with doing this and it doesn't cause any harm to the parser; it could just be a source of confusion to users. Rather than preventing this for no good reason, I think I prefer "return a reference to the value" approach... Unless someone has a compelling reason to know the index of a positional argument in the arg list... Anyone?

Reply:
The "required arg after an optional arg" edge case just seems messy to deal with, but I think returning a reference is probably the most elegant solution.

Reply:
👍 I will make this change.